### PR TITLE
[MXNET-96] Language model with Google's billion words dataset

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
This example reproduces the result (~42 perplexity) on [Exploring the Limits of Language Modeling](https://arxiv.org/pdf/1602.02410.pdf) on the GBW dataset. 
See `readme.mk` for details. 
@mli @szha @zheng-da @piiswrong 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/fully_connected.cc [hunk 1]
```diff
@@ -214,9 +238,7 @@ If ``no_bias`` is set to be true, then the ``bias`` term is ignored.
 .set_attr<nnvm::FInferShape>("FInferShape", FullyConnectedShape)
 .set_attr<nnvm::FInferType>("FInferType", FullyConnectedType)
 .set_attr<FCompute>("FCompute<cpu>", FullyConnectedCompute<cpu>)
-#if MXNET_USE_MKLDNN == 1
```
Comment (TaoLv) [inline src/operator/nn/fully_connected.cc]: Why remove this?

File: src/operator/nn/fully_connected.cc [hunk 2]
```diff
@@ -87,8 +90,16 @@ void FullyConnectedComputeExCPU(const nnvm::NodeAttrs& attrs,
     return;
   }
   FallBackCompute(FullyConnectedCompute<cpu>, attrs, ctx, inputs, req, outputs);
+#else
+  std::vector<TBlob> in_blobs(inputs.size());
+  for (size_t i = 0; i < in_blobs.size(); i++) in_blobs[i] = inputs[i].data();
+  std::vector<TBlob> out_blobs(outputs.size());
+  for (size_t i = 0; i < out_blobs.size(); i++) out_blobs[i] = outputs[i].data();
+  FullyConnectedCompute<cpu>(attrs, ctx, in_blobs, req, out_blobs);
+#endif
```
Comment (TaoLv) [inline src/operator/nn/fully_connected.cc]: I think ‘FallBackCompute’ should be used to fall back the computation to original cpu implementation.

Comment (eric-haibin-lin) [reply src/operator/nn/fully_connected.cc]: This block will only be executed when MKL is absent

File: src/operator/nn/fully_connected.cc (hunk 1 above)
Comment (eric-haibin-lin) [reply src/operator/nn/fully_connected.cc]: This is for adding sparse support for FC. See line 213.

Comment (TaoLv) [reply src/operator/nn/fully_connected.cc]: I see. Both sparse and mkldnn support need run into this dispatch.

File: src/operator/nn/fully_connected.cc (hunk 2 above)
Comment (eric-haibin-lin) [reply src/operator/nn/fully_connected.cc]: When USE_MKL=1,
I wanted to simply use the non-MKL FCForward with 
```
data.data(),
weight.data(),
bias.data(),
```
assuming `data.data()` returns a TBlob with normal cpu layout even if data is in MKL layout. `weight.data()` and `bias.data()` should always return normal cpu layout if weight and bias are `row_sparse`. Please advice.

**Comment by eric-haibin-lin** [comment]
@zihaolucky

File: src/operator/nn/fully_connected.cc [hunk 3]
```diff
@@ -73,22 +76,47 @@ static bool FullyConnectedShape(const nnvm::NodeAttrs& attrs,
...
+    } else {
+      FallBackCompute(FullyConnectedCompute<cpu>, attrs, ctx, inputs, req, outputs);
+    }
     return;
   }
-  FallBackCompute(FullyConnectedCompute<cpu>, attrs, ctx, inputs, req, outputs);
+#endif
+  if (valid_data && valid_weight && valid_bias && valid_out) {
+    std::vector<TBlob> in_blobs(inputs.size());
+    for (size_t i = 0; i < in_blobs.size(); i++) in_blobs[i] = inputs[i].data();
```
Comment (zheng-da) [inline src/operator/nn/fully_connected.cc]: Actually, if an NDArray uses MKLDNN format, data() will convert its layout inside the array. This caused a race condition, I just fixed. You shouldn't call data() on an NDArray with MKLDNN format. Please check FallBackCompute to see how it is handled correctly.

I also don't understand what happens if the input array is a row sparse. If the row sparse array doesn't have zero-entry rows, it works fine. What is the row sparse array have zero-entry rows, isn't the memory returned from data() is smaller than we expect?

**Comment by sxjscience** [comment]
@eric-haibin-lin The RNN.unroll is just updated to support variable length. Need to resolve the conflict.

File: src/operator/nn/fully_connected.cc (hunk 3 above)
Comment (pengzhao-intel) [reply src/operator/nn/fully_connected.cc]: So, the `data()` should be improved functionality. The developer doesn't know there's a potential issue when working with other formats of NDArray.

Comment (zheng-da) [reply src/operator/nn/fully_connected.cc]: I guess the only option is to disable data() for NDArrays with MKLDNN format.

File: python/mxnet/gluon/contrib/rnn/rnn_cell.py [hunk 4]
```diff
@@ -181,3 +181,126 @@ def unroll(self, length, inputs, begin_state=None, layout='NTC', merge_outputs=N
         outputs, _, _, _ = _format_sequence(length, outputs, layout, merge_outputs)
 
         return outputs, states
+
+
+class LSTMPCell(HybridRecurrentCell):
```
Comment (sxjscience) [inline python/mxnet/gluon/contrib/rnn/rnn_cell.py]: Can this be implemented as a ModifierCell? The implementation should be similar as ZoneOut.

File: src/operator/nn/fully_connected.cc [hunk 5]
```diff
@@ -56,7 +56,10 @@ static bool FullyConnectedShape(const nnvm::NodeAttrs& attrs,
   }
   SHAPE_ASSIGN_CHECK(*in_shape, fullc::kWeight, Shape2(param.num_hidden, num_input));
   if (!param.no_bias) {
-    SHAPE_ASSIGN_CHECK(*in_shape, fullc::kBias, Shape1(param.num_hidden));
+    if (!shape_assign(&(*in_shape)[fullc::kBias], Shape1(param.num_hidden)) &&
+        !shape_assign(&(*in_shape)[fullc::kBias], Shape2(param.num_hidden, 1))) {
```
Comment (sxjscience) [inline src/operator/nn/fully_connected.cc]: Why should we check it against (num_hidden, 1).

File: example/rnn/large_word_lm/data.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                word, count = line.strip().split()
+                vocab.add(word, int(count))
+        vocab.finalize()
+        return vocab
+
+class Dataset(object):
+    """ A dataset for truncated bptt with multiple sentences.
+        Adapeted from @rafaljozefowicz's implementation.
+     """
+    def __init__(self, vocab, file_pattern, deterministic=False):
```
Comment (szha) [inline example/rnn/large_word_lm/data.py]: `shuffle` instead of `deterministic`, since a fixed random seed may still produce deterministic but shuffled results.

File: example/rnn/large_word_lm/model.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+# under the License.
+
+import mxnet as mx
+import mxnet.symbol as S
+import numpy as np
+
+def cross_entropy_loss(inputs, labels, rescale_loss=1):
+    """ cross entropy loss """
+    criterion = mx.gluon.loss.SoftmaxCrossEntropyLoss()
+    loss = criterion.hybrid_forward(S, inputs, labels)
```
Comment (szha) [inline example/rnn/large_word_lm/model.py]: loss = criterion(inputs, labels) should do. Also, `rescale_loss` can be put in the constructor call of the SoftmaxCELoss using argument `weight`

File: python/mxnet/gluon/contrib/rnn/rnn_cell.py (hunk 4 above)
Comment (szha) [reply python/mxnet/gluon/contrib/rnn/rnn_cell.py]: Good idea

File: src/operator/nn/fully_connected-inl.h [hunk 8]
```diff
@@ -95,11 +95,16 @@ void FCForward(const OpContext &ctx, const FullyConnectedParam &param,
...
+    << "weight.data().shape[1] != data.data().shape[1]. Not supported by FCForward";
   // Legacy approach shown here for comparison:
   //   out = dot(data, wmat.T());
   linalg_gemm(data, wmat, out, false, true, s);
   if (!param.no_bias) {
-    Tensor<xpu, 1, DType> bias = in_data[fullc::kBias].get<xpu, 1, DType>(s);
+    Tensor<xpu, 1, DType> bias = in_data[fullc::kBias].get_with_shape<xpu, 1, DType>(
+      Shape1(wmat.shape_[0]), s);
+    CHECK_EQ(bias.shape_[0], wmat.shape_[0])
+      << "bias.data().shape[0] != weight.data().shape[0]. Not supported by FCForward";
```
Comment (szha) [inline src/operator/nn/fully_connected-inl.h]: Use words to describe the error instead. Also, if `flatten is True` the error message for `data.data()` might not make sense.

File: example/rnn/large_word_lm/data.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+import numpy as np
+import codecs, glob, random, logging
+
+class Vocabulary(object):
+    """ A dictionary for words.
+        Adapeted from @rafaljozefowicz's implementation.
+    """
+    def __init__(self):
+        self._token_to_id = {}
+        self._token_to_count = {}
```
Comment (szha) [inline example/rnn/large_word_lm/data.py]: collections.Counter?

File: python/mxnet/gluon/contrib/rnn/rnn_cell.py (hunk 4 above)
Comment (eric-haibin-lin) [reply python/mxnet/gluon/contrib/rnn/rnn_cell.py]: Is there an easy way to modify state in Modifier cell (especially during unroll)??

File: src/operator/nn/fully_connected.cc (hunk 5 above)
Comment (eric-haibin-lin) [reply src/operator/nn/fully_connected.cc]: The bias are trained with sparse embedding with requires a 2-D shape.

File: example/rnn/large_word_lm/data.py (hunk 6 above)
Comment (eric-haibin-lin) [reply example/rnn/large_word_lm/data.py]: ok

File: example/rnn/large_word_lm/model.py (hunk 7 above)
Comment (eric-haibin-lin) [reply example/rnn/large_word_lm/model.py]: ok

File: example/rnn/large_word_lm/data.py (hunk 9 above)
Comment (eric-haibin-lin) [reply example/rnn/large_word_lm/data.py]: ok

File: python/mxnet/gluon/contrib/rnn/rnn_cell.py (hunk 4 above)
Comment (eric-haibin-lin) [reply python/mxnet/gluon/contrib/rnn/rnn_cell.py]: I don't it's a good abstraction as a modifier cell. The projection only happens on the hidden state of LSTM, not on the cell state. I am not sure what's the expected behavior for GRU cell and RNN cell where the number of states is just one. Also the paper call this LSTMP, it's specific to LSTM. I think it's find to just call it `contrib.LSTMP` which inherits `HybridRecurrentCell `.

**Comment by eric-haibin-lin** [comment]
@TaoLv  @zheng-da could you review the updated code for FC?

File: src/operator/nn/fully_connected.cc [hunk 10]
```diff
@@ -73,22 +76,61 @@ static bool FullyConnectedShape(const nnvm::NodeAttrs& attrs,
...
+                         outputs);
+    } else {
+      FallBackCompute(FullyConnectedCompute<cpu>, attrs, ctx, inputs, req, outputs);
+    }
     return;
+  } else if (valid_data && valid_weight && valid_bias && valid_out) {
+    // inputs
+    std::vector<TBlob> in_blobs(inputs.size());
+    auto get_data = [](const NDArray& nd) -> TBlob {
+      if (nd.storage_type() == kDefaultStorage) return nd.Reorder2Default().data();
```
Comment (zheng-da) [inline src/operator/nn/fully_connected.cc]: nd.Reorder2Default() returns a temp object here, which controls a piece of memory. This memory will be freed when the function returns, and TBlob will reference to a free'd memory.

**Comment by eric-haibin-lin** [comment]
Anything else to address? 
@sxjscience @zheng-da @szha @cjolivier01 @TaoLv

**Comment by zheng-da** [comment]
looks good to me
